### PR TITLE
V.3.0.0

### PR DESCRIPTION
**Pull request by v-ruboranov**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # powerbi-visuals-gantt
-[![Build Status](https://travis-ci.org/Microsoft/powerbi-visuals-gantt.svg?branch=master)](https://travis-ci.org/Microsoft/powerbi-visuals-gantt) [![Coverage Status](https://coveralls.io/repos/github/Microsoft/powerbi-visuals-gantt/badge.svg?branch=master)](https://coveralls.io/github/Microsoft/powerbi-visuals-gantt?branch=master)
+[![Build Status](https://travis-ci.org/Microsoft/powerbi-visuals-gantt.svg?branch=master)](https://travis-ci.org/Microsoft/powerbi-visuals-gantt)
```
Comment (snapp1) [inline README.md:2]: image

build status is not shown properly, remove it from Readme

File: test/visualData.ts [hunk 2]
```diff
@@ -102,7 +102,33 @@ export class VisualData extends TestDataViewBuilder {
         return result;
     }
 
-    public getDataView(columnNames?: string[], withMilestones?: boolean): DataView {
+    public generateHightLightedValues(length: number, hightlightedElementNumber?: number): number[] {
```
Comment (snapp1) [inline test/visualData.ts:105]: typo HighLighted

File: test/visualTest.ts [hunk 3]
```diff
@@ -303,51 +310,51 @@ describe("Gantt", () => {
...
                 expect(progressOfTasks.length).toEqual(returnTasks.length);
                 done();
             });
         });
 
-        it("Task Completion width is equal task width", (done) => {
-            defaultDataViewBuilder.valuesCompletePrecntege = VisualData.getRandomUniqueNumbers(
-                defaultDataViewBuilder.valuesTaskTypeResource.length, 0, 100
-            );
+        // it("Task Completion width is equal task width", (done) => {
```
Comment (snapp1) [inline test/visualTest.ts:321]: why commented?

File: test/visualTest.ts [hunk 4]
```diff
@@ -1712,32 +1730,31 @@ describe("Gantt", () => {
...
-                    visualBuilder.taskRect.toArray().map($).forEach((e, i) => {
-                        // if completion is null (no info about completion) - task expected to be completed
-                        const expectedOpacity = tasks[i].completion ? VisualClass["NotCompletedTaskOpacity"].toString() : VisualClass["TaskOpacity"].toString();
-                        expect(e.css("opacity")).toBe(expectedOpacity);
-                    });
-                    done();
-                });
-            });
-        });
+        // describe("Task Completion", () => {
```
Comment (snapp1) [inline test/visualTest.ts:1733]: why commented?

File: test/visualTest.ts [hunk 5]
```diff
@@ -2081,4 +2105,181 @@ describe("Gantt", () => {
...
+            };
+
+
+            const promise1 = new Promise((resolve)=> {
+                setTimeout(() => {
+                    dataView.metadata.objects = objects1;
+                    visualBuilder.update(dataView);
+                    resolve(visualBuilder.instance[collapsedTasksUpdateIDs].includes(collapsedTasksUpdatIDsRandom[0]));
+                }, 
+                1 * 1000);
```
Comment (snapp1) [inline test/visualTest.ts:2259]: why 1 * 1000?

File: src/gantt.ts [hunk 6]
```diff
@@ -1395,20 +1428,25 @@ export class Gantt implements IVisual {
     * Gets all unique types from the tasks array
     * @param dataView The data model.
     */
-    private static getAllTasksTypes(dataView: DataView): TaskTypes {
+    private getAllTasksTypes(dataView: DataView): TaskTypes {
```
[the remark this answers is not in the document]
Comment (v-ruboranov) [reply src/gantt.ts]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/c6964bc5dc187032882e5e618d11f6ca46fdd8c7 fixed here

File: src/gantt.ts [hunk 7]
```diff
@@ -860,20 +884,22 @@ export class Gantt implements IVisual {
...
+        // eslint-disable-next-line max-lines-per-function
         values.Task.forEach((categoryValue: PrimitiveValue, index: number) => {
             let color: string = taskColor || Gantt.DefaultValues.TaskColor;
             let completion: number = 0;
             let taskType: TaskTypeMetadata = null;
             let wasDowngradeDurationUnit: boolean = false;
-            let tooltips: VisualTooltipDataItem[] = [];
+            const tooltips: VisualTooltipDataItem[] = [];
             let stepDurationTransformation: number = 0;
+            let highlight: number = null;
```
[the remark this answers is not in the document]
Comment (v-ruboranov) [reply src/gantt.ts]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/c6964bc5dc187032882e5e618d11f6ca46fdd8c7 fixed here

File: src/gantt.ts [hunk 8]
```diff
@@ -670,19 +693,17 @@ export class Gantt implements IVisual {
      * @param dataView The data Model
      * @param cultureSelector The current user culture
      */
-    private static getFormatters(
+    private getFormatters(
```
[the remark this answers is not in the document]
Comment (v-ruboranov) [reply src/gantt.ts]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/c6964bc5dc187032882e5e618d11f6ca46fdd8c7 restored here

File: test/visualData.ts (hunk 2 above)
Comment (v-ruboranov) [reply test/visualData.ts]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/0557ec67b0fcaa516eb70ffbb4a2e0d3a903ccea resolved here

File: src/gantt.ts [hunk 9]
```diff
@@ -1395,20 +1428,25 @@ export class Gantt implements IVisual {
...
     * @param dataView The data model.
     */
-    private static getAllTasksTypes(dataView: DataView): TaskTypes {
+    private getAllTasksTypes(dataView: DataView): TaskTypes {
         const taskTypes: TaskTypes = {
             typeName: "",
             types: []
         };
-        let index: number = _.findIndex(dataView.metadata.columns, col => col.roles.hasOwnProperty(GanttRoles.Legend));
+        const index: number = _.findIndex(dataView.metadata.columns, col => Object.prototype.hasOwnProperty.call(col.roles, GanttRoles.Legend));
```
[the remark this answers is not in the document]
Comment (v-ruboranov) [reply src/gantt.ts]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/52274fe5a79c5ecef1dea7f94c4d328cbc2f2008 refactored here

File: src/gantt.ts [hunk 10]
```diff
@@ -1309,41 +1341,39 @@ export class Gantt implements IVisual {
...
+        const settings: GanttChartSettingsModel = this.parseSettings(dataView, colorHelper);
+        
+        const taskTypes: TaskTypes = this.getAllTasksTypes(dataView);
         const formatters: GanttChartFormatters = this.getFormatters(dataView, settings, host.locale || null);
 
-        const isDurationFilled: boolean = _.findIndex(dataView.metadata.columns, col => col.roles.hasOwnProperty(GanttRoles.Duration)) !== -1,
-            isEndDateFillled: boolean = _.findIndex(dataView.metadata.columns, col => col.roles.hasOwnProperty(GanttRoles.EndDate)) !== -1,
-            isParentFilled: boolean = _.findIndex(dataView.metadata.columns, col => col.roles.hasOwnProperty(GanttRoles.Parent)) !== -1,
-            isResourcesFilled: boolean = _.findIndex(dataView.metadata.columns, col => col.roles.hasOwnProperty(GanttRoles.Resource)) !== -1;
+        const isDurationFilled: boolean = _.findIndex(dataView.metadata.columns, col => Object.prototype.hasOwnProperty.call(col.roles, GanttRoles.Duration)) !== -1,
```
[the remark this answers is not in the document]
Comment (v-ruboranov) [reply src/gantt.ts]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/52274fe5a79c5ecef1dea7f94c4d328cbc2f2008 refactored here

File: package.json [hunk 11]
```diff
@@ -28,57 +28,93 @@
...
+    "less": "^4.1.3",
+    "less-loader": "^11.1.0",
+    "playwright": "^1.33.0",
+    "powerbi-models": "^1.12.6",
+    "powerbi-visuals-api": "~5.4.0",
+    "powerbi-visuals-tools": "^4.3.1",
+    "powerbi-visuals-utils-testutils": "^6.0.1",
+    "style-loader": "^3.3.2",
+    "ts-loader": "^9.4.2",
+    "ts-node": "^10.9.1",
```
[the remark this answers is not in the document]
Comment (v-ruboranov) [reply package.json]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/f078df1aa942f1267f0d5814297430a9783a5f29 removed ts-node package in this commit

File: package.json [hunk 12]
```diff
@@ -28,57 +28,93 @@
...
-    "lodash": "^4.17.15",
-    "powerbi-visuals-utils-chartutils": "2.4.0",
-    "powerbi-visuals-utils-colorutils": "2.2.0",
-    "powerbi-visuals-utils-dataviewutils": "2.2.0",
-    "powerbi-visuals-utils-formattingutils": "^4.4.1",
-    "powerbi-visuals-utils-interactivityutils": "5.5.2",
-    "powerbi-visuals-utils-svgutils": "2.2.0",
-    "powerbi-visuals-utils-tooltiputils": "^2.3.0",
-    "powerbi-visuals-utils-typeutils": "2.2.0"
+    "@typescript-eslint/eslint-plugin": "^5.59.2",
```
[the remark this answers is not in the document]
Comment (v-ruboranov) [reply package.json]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/28fd84fe5a56dfeb5431ac89656f0ab656eff9cc resolved here

File: src/columns.ts [hunk 13]
```diff
@@ -37,32 +38,37 @@ import PrimitiveValue = powerbi.PrimitiveValue;
...
     public static getGroupedValueColumns(dataView: DataView): GanttColumns<DataViewValueColumn>[] {
-        let categorical: DataViewCategorical = dataView && dataView.categorical;
-        let values: DataViewValueColumns = categorical && categorical.values;
-        let grouped: DataViewValueColumnGroup[] = values && values.grouped();
-        return grouped && grouped.map(g => _.mapValues(
+        const categorical: DataViewCategorical = dataView?.categorical;
+        const values: DataViewValueColumns = categorical?.values;
+        const grouped: DataViewValueColumnGroup[] = values?.length && values.grouped();
+
+        if (values === undefined || values.length == 0) {
```
[the remark this answers is not in the document]
Comment (v-ruboranov) [reply src/columns.ts:53]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/193503e61d45e5195b9783ed84609488143b424d resolved here

File: src/gantt.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (v-ruboranov) [reply src/gantt.ts]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/622f2eb3cee597e5a98115d133cd5cf93ecc4228 refactored here

File: src/gantt.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (v-ruboranov) [reply src/gantt.ts]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/622f2eb3cee597e5a98115d133cd5cf93ecc4228 refactored here

File: src/gantt.ts [hunk 14]
```diff
@@ -374,66 +386,78 @@ export class Gantt implements IVisual {
...
+    }
+
     /**
      * Create the viewport area of the gantt chart
      */
     private createViewport(element: HTMLElement): void {
-        let self = this;
-        const isHighContrast: boolean = this.colorHelper.isHighContrast;
+        // eslint-disable-next-line
+        const self: Gantt = this;
```
[the remark this answers is not in the document]
Comment (v-ruboranov) [reply src/gantt.ts]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/066198a5368b31c01ce874e7348504c87df02ef7 fixed here

File: src/gantt.ts [hunk 15]
```diff
@@ -2617,15 +2684,16 @@ export class Gantt implements IVisual {
                 .style("fill", taskResourceColor)
                 .style("font-size", PixelConverter.fromPoint(taskResourceFontSize));
 
-            let self: Gantt = this;
-            let hasNotNullableDates: boolean = this.hasNotNullableDates;
+            // eslint-disable-next-line
+            const self: Gantt = this;
```
[the remark this answers is not in the document]
Comment (v-ruboranov) [reply src/gantt.ts]: https://github.com/microsoft/powerbi-visuals-gantt/pull/296/commits/31dcf6234903296077cf937453bf15c3a924c181 resolved here

**Pull request closed without merging**